File: src/main/java/net/focik/homeoffice/finance/domain/fee/Fee.java
package net.focik.homeoffice.finance.domain.fee;

import lombok.*;
import net.focik.homeoffice.finance.domain.firm.Firm;
import net.focik.homeoffice.finance.domain.payment.FinancialTransaction;
import net.focik.homeoffice.finance.domain.payment.Payment;
import net.focik.homeoffice.utils.share.PaymentStatus;
import net.focik.homeoffice.utils.share.PaymentType;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

@Builder
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@ToString
public class Fee implements FinancialTransaction {

    private int id;
    private Firm firm;
    private int idUser;
    private String name;
    private String feeNumber;
    private LocalDate date;
    private FeeFrequencyEnum feeFrequency;
    private Integer numberOfPayments;//ile razy pobrać
    private BigDecimal amount;
    private LocalDate firstPaymentDate;
    private String accountNumber;
    private PaymentStatus feeStatus;
    private String otherInfo;
    private List<FeeInstallment> installments;

    public void addFeeInstallment(List<FeeInstallment> feeInstallments) {
        this.installments = feeInstallments;
    }

    public void changeFeeStatus(PaymentStatus feeStatus) {
        this.feeStatus = feeStatus;
    }

    @Override
    public Payment getPayment() {
        return Payment.builder()
                .id(id)
                .idUser(idUser)
                .name(name)
                .paymentDay(firstPaymentDate.getDayOfMonth())
                .paymentType(PaymentType.FEE)
                .installments(installments)
                .build();
    }
}

File: src/types/Fee.ts
import type { Firm } from './Firm'
import type {PaymentStatus} from "./Payment.ts";

export interface Fee {
  id: number
  firm: Firm | null
  idUser: number
  name: string
  feeNumber: string
  date: Date | null
  feeFrequency: FeeFrequency | null
  numberOfPayments: number
  amount: number
  firstPaymentDate: Date | null
  accountNumber: string
  feeStatus: PaymentStatus
  otherInfo: string
  installmentList: FeeInstallment[]
}

export interface FeeInstallment {
  idFeeInstallment: number
  idFee: number
  installmentAmountToPay: number
  installmentAmountPaid: number
  paymentDeadline: Date | null
  paymentDate: Date | null
  paymentStatus: PaymentStatus
}

export interface FeeFrequency {
  name: string
  viewName: string
  frequencyNumber: number
}
